Lib/typing.py

```diff
@@ -1798,7 +1798,13 @@ def __instancecheck__(cls, instance):
             raise TypeError("Instance and class checks can only be used with"
                             " @runtime_checkable protocols")
 
-        if super().__instancecheck__(instance):
+        # gh-104555: Don't call super().__instancecheck__ here,
+        # ABCMeta.__instancecheck__ would erroneously use it to populate the cache,
+        # which would cause incorrect results for *issubclass()* calls
+        if type.__instancecheck__(cls, instance):
+            return True
+
+        if cls.__callable_proto_members_only__ and issubclass(type(instance), cls):
             return True
```

Comment:
Without this check, I realised that this patch would also have slowed down `isinstance(3, SupportsIndex)` quite dramatically.